source/LiteralModel.cpp

```diff
@@ -0,0 +1,117 @@
+  static void raise(const std::string& what) {
+    ERROR(1, "Literal Model Consistency Error: {}", what);
+    EventLogger::log_event("regex_model_consistency_error", what);
+  }
+};
+
+}
+
+static void check_taint_config_consistency(const std::string& pattern, const TaintConfig& frame) {
+  if (frame.kind() == nullptr) {
+    LiteralModelConsistencyError::raise(fmt::format(
+        "Model for regex `{}` must have a source kind.", pattern));
+  }
+}
+
+static Taint create_taint(const std::string& pattern, const std::vector<TaintConfig>& sources) {
+  Taint all_sources;
+  for (const auto& source : sources) {
+    mt_assert(source.is_leaf());
+    check_taint_config_consistency(pattern, source);
+    all_sources.join_with(Taint{std::move(source)});
+  }
+  return all_sources;
+}
+
+LiteralModel::LiteralModel() : regex_(pattern_) { }
+
+LiteralModel::LiteralModel(
+    std::string&& pattern,
+    const std::vector<TaintConfig>& sources)
+    : pattern_(std::move(pattern)), regex_(pattern_), sources_(create_taint(pattern_, sources)) {
+}
+
+LiteralModel::LiteralModel(const LiteralModel& other)
+    : pattern_(other.pattern_), regex_(pattern_), sources_(other.sources_) {
+}
+
+const std::string& LiteralModel::pattern() const {
+  return pattern_;
+}
+
+const Taint& LiteralModel::sources() const {
+  return sources_;
+}
+
+bool LiteralModel::matches(const std::string_view literal) const {
+  return re2::RE2::FullMatch(literal, regex_);
+}
+
+void LiteralModel::join_with(const LiteralModel& other) {
+  if (this != &other) {
+    sources_.join_with(other.sources_);
+  }
+}
+
+bool LiteralModel::empty() const {
+  return sources_.is_bottom();
+}
+
+LiteralModel LiteralModel::from_json(
+    const Json::Value& value,
+    Context& context) {
+  JsonValidation::check_unexpected_members(
+      value, {"pattern", "sources", "description"});
+
+  std::vector<TaintConfig> sources;
+  for (auto source : JsonValidation::null_or_array(value, /* field */ "sources")) {
+    TaintConfig taintConfig{TaintConfig::from_json(source, context)};
+    if (!taintConfig.is_leaf())
+      throw JsonValidationError{source, /* field */ "sources", "empty callee"};
+
+    sources.push_back(std::move(taintConfig));
+  }
+  return LiteralModel(JsonValidation::string(value, "pattern"), sources);
+}
+
+Json::Value LiteralModel::to_json(const ExportOriginsMode export_origins_mode) const {
+  auto value = Json::Value(Json::objectValue);
+
+  value["pattern"] = pattern_;
+
+  if (!sources_.is_bottom()) {
+    auto sources_value = Json::Value(Json::arrayValue);
+    for (const auto& source : sources_.frames_iterator()) {
+      mt_assert(!source.is_bottom());
+      sources_value.append(source.to_json(export_origins_mode));
+    }
+    value["sources"] = sources_value;
+  }
+
+  return value;
+}
+
+Json::Value LiteralModel::to_json(Context& context) const {
+  auto value = to_json(context.options->export_origins_mode());
+  value["position"] = context.positions->unknown()->to_json();
```

Comment:
Not sure I understand why we want to export the empty position? It's just wasting space?

Reply:
As with a few other items, that's because I started from `FieldModel` and that's how it behaves:
https://github.com/facebook/mariana-trench/blob/f5c7c535f30ff554e8c5643988535f0736577d4e/source/FieldModel.cpp#L188

Happy to delete this line though. 🙂

Reply:
Fine, let's keep it for consistency.